### PR TITLE
Set the mouse cursor in the canvas based on the current tool and its state

### LINKED ISSUE #463: Set mouse cursor in the viewport based on tool state
Add a FrontendMessage to set the cursor to an enum value with variants like `Default`, `ZoomIn`, `ZoomOut`, `Grabbing`, and `Crosshair`. Then apply the [CSS `cursor` rule](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor) and have the document viewport `<div class="canvas" ref="canvas">` apply the cursor style property like `<div class="canvas" ref="canvas" :style="{ cursor: canvasCursor }">` where the default cursor is the JS `undefined` value (so Vue applies no style value).

The Navigate tool should display `ZoomIn` by default since <kbd>LMB</kbd> clicking causes a zoom in operation, and <kbd>LMB</kbd> dragging zooms in and out. While holding the <kbd>Shift</kbd> modifier, switch to the `ZoomOut` cursor since <kbd>LMB</kbd> clicking with that held down zooms out. When <kbd>MMB</kbd> dragging, switch to the `Grabbing` cursor (both in the Navigate tool and the global behavior). For rotating the document, that can stay the default cursor until we design a custom SVG cursor for that (and probably all icons but the default cursor icon).

The shape drawing tools (Line, Rectangle, Ellipse, and Shape) should all get the `Crosshair` cursor.

The Select, Path, Pen, Eyedropper, Fill, and other future tools will eventually get their own SVG icons also. @Keavon needs to design those in a format that's optimized for a mouse cursor.

Ideally, the cursor should stay the same when the mouse leaves the viewport `.canvas` element while a drag is occurring. [See this](https://developer.mozilla.org/en-US/docs/Web/API/Element/setPointerCapture) for how.

### PR DESCRIPTION
**Pull request by asyncth**
Closes #463

### CODE REVIEW

**Comment by asyncth** [comment]
Forgot to add the `setCapture` thing, will do tomorrow.

File: editor/src/frontend/utility_types.rs [hunk 1]
```diff
@@ -6,3 +6,12 @@ pub struct FrontendDocumentDetails {
 	pub name: String,
 	pub id: u64,
 }
+
+#[derive(Clone, Copy, Debug, Eq, Deserialize, PartialEq, Serialize)]
+pub enum FrontendMouseCursor {
```
Comment (Keavon) [inline editor/src/frontend/utility_types.rs:11]: "Frontend" is perhaps redundant here. Maybe `MouseCursorIcon` would be more descriptive.

File: frontend/src/components/panels/Document.vue [hunk 2]
```diff
@@ -123,7 +123,7 @@
 						<CanvasRuler :origin="rulerOrigin.y" :majorMarkSpacing="rulerSpacing" :numberInterval="rulerInterval" :direction="'Vertical'" />
 					</LayoutCol>
 					<LayoutCol :class="'canvas-area'">
-						<div class="canvas" ref="canvas">
+						<div class="canvas" ref="canvas" :style="{ cursor: canvasCursor }" v-on:pointerdown="canvasPointerDown">
```
Comment (Keavon) [inline frontend/src/components/panels/Document.vue:126]: `@pointerdown="(e: PointerEvent) => canvasPointerDown(e)"`

File: frontend/src/components/panels/Document.vue [hunk 3]
```diff
@@ -338,6 +339,10 @@ export default defineComponent({
 		resetWorkingColors() {
 			this.editor.instance.reset_colors();
 		},
+		canvasPointerDown(event: PointerEvent) {
```
Comment (Keavon) [inline frontend/src/components/panels/Document.vue:342]: `e`, not `event`, is convention.

File: frontend/src/components/panels/Document.vue (hunk 2 above)
Comment (asyncth) [reply frontend/src/components/panels/Document.vue:126]: What's wrong with `@pointerdown="canvasPointerDown"`?

File: frontend/src/dispatcher/js-messages.ts [hunk 4]
```diff
@@ -39,6 +39,8 @@ export class FrontendDocumentDetails extends DocumentDetails {
 	readonly id!: BigInt;
 }
 
+export type MouseCursorIcon = string;
```
Comment (Keavon) [inline frontend/src/dispatcher/js-messages.ts:42]: Change the type from `string` to `"default" | "zoom-in" | "zoom-out" | "grabbing" | "crosshair"`

File: frontend/src/dispatcher/js-messages.ts [hunk 5]
```diff
@@ -201,6 +203,33 @@ export class UpdateDocumentRulers extends JsMessage {
...
+
+		case "Grabbing":
+			return "grabbing";
+
+		case "Crosshair":
+			return "crosshair";
+
+		default:
+			return "default";
+	}
```
Comment (Keavon) [inline frontend/src/dispatcher/js-messages.ts:225]: A dictionary lookup would be cleaner here:
```ts
const cssNames = {
    ZoomIn: "zoom-in",
    ZoomOut: "zoom-out",
    Grabbing: "grabbing",
    Crosshair: "crosshair",
};
return cssNames[value] || "default";
```

File: frontend/src/dispatcher/js-messages.ts (hunk 4 above)
Comment (Keavon) [reply frontend/src/dispatcher/js-messages.ts]: And please move this down next to where it's used. Right above its first usage (in the `Transform(...)` would be best.

File: frontend/src/components/panels/Document.vue (hunk 2 above)
Comment (Keavon) [reply frontend/src/components/panels/Document.vue]: We're being explicit with our type usage by convention.

**Review by Keavon — APPROVED** [review]

**Pull request merged**